hbase-server/src/main/java/org/apache/hadoop/hbase/master/ServerManager.java

```diff
@@ -324,8 +324,19 @@ public void regionServerReport(ServerName sn, ServerMetrics sl) throws YouAreDea
       // the ServerName to use. Here we presume a master has already done
       // that so we'll press on with whatever it gave us for ServerName.
       if (!checkAndRecordNewServer(sn, sl)) {
-        LOG.info("RegionServerReport ignored, could not record the server: " + sn);
-        return; // Not recorded, so no need to move on
+        // Master already registered server with same (host + port) and higher startcode.
```

Comment:
I still do not think this is necessary, because if the new server with the same host and port has already registered to master, how can we return this `YouAreDeadException` to the old server? Even if there is a race condition, when sending we will receive a connection reset because the old server is already dead...

Reply:
When it happened (as per logs mentioned on the jira), master processed the report and that generated inconsistencies.

We have seen this happen many times in the past when regionserver is not really aborted but looses connection with Zookeeper, triggering SCP by master. And regionserver with new startcode is not only alive but has also reported regionservers to master. After that, somehow master still receives regionserver report, master processes it and that results into inconsistencies. I know this is rare case but it definitely happened more than once in more than one prod clusters.

Reply:
This is more of safety check, it will prevent inconsistencies. I agree that anyone looking at this would think, why do we need such extra safety, it's valid point but I can guarantee that not having such strict validation has caused inconsistencies.

Reply:
The other way to think about this is: why should we even receive any report from old server and not throw YouAreDeadException while we already know that new server is alive and is already registered? 🙂

Reply:
I checked the code, now I understand why we need to throw an exception here. Your comment totally missed the most important part...

At least your comment should include these two points:

1. The exception thrown here is not meant to tell the region server it is dead because if there is a new server on the same host port, the old server should have already been dead.
2. The exception thrown here is to skip the later steps of the whole regionServerReport request processing. Usually, after recording it in ServerManager, we will call the related methods in AssignmentManager to record region states. If the region server is already dead, we should not do these steps any more, so here we throw an exception to let the upper layer know that they should not continue processing any more.

Reply:
Sounds good, will add more comments to make it clear. Thanks Duo!